src/datadog/trace_segment.cpp

```diff
@@ -245,6 +248,43 @@ void TraceSegment::span_finished() {
     span.tags[tags::internal::runtime_id] = runtime_id_.string();
   }
 
+  // calculate http.endpoint if:
+  // a) the feature is not disabled, and
+  // b) the tag http.endpoint is not already set, and
+  // c) http.url is set, and
+  // d) http.route is not set or resource_renaming_mode is ALWAYS_CALCULATE
+  if (resource_renaming_mode_ != ResourceRenamingMode::DISABLED &&
+      local_root.tags.find(tags::internal::http_endpoint) ==
+          local_root.tags.end()) {
+    auto http_url_tag = local_root.tags.find(tags::internal::http_url);
+    const bool should_calculate_endpoint =
+        http_url_tag != local_root.tags.end() &&
+        (resource_renaming_mode_ == ResourceRenamingMode::ALWAYS_CALCULATE ||
+         local_root.tags.find(tags::internal::http_route) ==
+             local_root.tags.end());
+
+    if (should_calculate_endpoint) {
+      // extract just the path from the http.url for endpoint guessing
+      // http.url format is: scheme://host:port/path?query
+      std::string_view path = http_url_tag->second;
+
+      auto scheme_end = path.find("://");
+      if (scheme_end != std::string_view::npos) {
+        path.remove_prefix(scheme_end + 3);  // skip scheme and ://
+        auto path_start = path.find('/');
+        if (path_start != std::string_view::npos) {
+          path.remove_prefix(path_start);
+        } else {
+          path = "/";
+        }
```

Comment:
I’d suggest moving this logic into a separate function. Given the amount of URL handling, it also makes sense to follow up on my earlier comment about reusing the URL parser.